src/fsharp/FSharp.Core/map.fs

```diff
@@ -598,6 +613,56 @@ namespace Microsoft.FSharp.Collections
             member s.TryGetValue(k,r) = if s.ContainsKey(k) then (r <- s.[k]; true) else false
             member s.Remove(k : 'Key) = ignore(k); (raise (NotSupportedException(SR.GetString(SR.mapCannotBeMutated))) : bool)
 
+        interface System.Collections.IDictionary with
+            member m.Item
+                with get key =
+                    match key with
+                    | null ->
+                        // According to the documentation for IDictionary, implementations should
+                        // raise ArgumentNullException if the key passed to the indexer is null.
+                        nullArg "key"
+                    | :? 'Key as key ->
+                        match m.TryFind key with
+                        | Some v -> box v
+                        | None -> null
+                    | _ -> null
+                and  set key value =
```

Comment:
I have an emotional problem with these kind of partial implementations of contracts.  it has always been one of my feefs with dotnet that it was so easy to do this kind of thing..  Throwing not implemented on an object that claims to implement a contract is .... well ... lame.  Having said that adding mutability to map would have been worse.

I really have to wonder what the use case for this feature is.  I could imagine implementing IReadOnlyDictionary and IReadOnlyCollection for interop purposes though.